crates/driver/src/domain/competition/auction.rs

```diff
@@ -76,12 +62,95 @@ impl Auction {
-    /// are more likely to be fulfilled come before less likely orders.
+    /// The orders for the auction.
     pub fn orders(&self) -> &[competition::Order] {
         &self.orders
     }
 
+    // Prioritize the orders such that those which are more likely to be fulfilled
+    // come before less likely orders. Filter out orders which the trader doesn't
+    // have enough balance to pay for.
+    //
+    // Prioritization is skipped during quoting. It's only used during competition.
+    pub async fn prioritize(mut self, eth: &Ethereum) -> Self {
+        // Sort orders so that most likely to be fulfilled come first.
+        self.orders.sort_by_key(|order| {
+            // Market orders are preferred over limit orders, as the expectation is that
+            // they should be immediately fulfillable. Liquidity orders come last, as they
+            // are the most niche and rarely used.
+            let class = match order.kind {
+                competition::order::Kind::Market => 2,
+                competition::order::Kind::Limit { .. } => 1,
+                competition::order::Kind::Liquidity => 0,
+            };
+            std::cmp::Reverse((
+                class,
+                // TODO No need to expose Tokens anymore, just use the auction
+                // If the orders are of the same kind, then sort by likelihood of fulfillment
+                // based on token prices.
+                order.likelihood(&self.tokens),
+            ))
+        });
+
+        // Fetch balances of each token for each trader.
+        let tokens_by_trader = self
+            .orders
+            .iter()
+            .flat_map(|order| {
+                [
+                    (order.trader(), order.sell.token),
+                    (order.trader(), order.buy.token),
+                ]
+            })
+            .unique()
+            .collect_vec();
+        let mut balances: HashMap<(order::Trader, eth::TokenAddress), eth::TokenAmount> =
+            join_all(tokens_by_trader.into_iter().map(|(trader, token)| {
+                async move {
+                    let balance = eth
+                        .balance_of(trader.into(), token)
+                        .await
+                        // If fetching the balance of this trader fails, don't filter out their
+                        // orders. Pretend like they have infinite balance.
+                        .unwrap_or(eth::U256::MAX.into());
+                    ((trader, token), balance)
+                }
+            }))
+            .await
+            .into_iter()
+            .collect();
+
+        // Filter out orders which the trader doesn't have enough balance to pay for.
+        self.orders.retain(|order| {
+            // Partial orders don't require the trader to hold the full balance, since they
+            // don't have to be fully fulfilled.
+            if order.is_partial() {
+                return true;
+            }
```

Comment:
The problem then is that the following can happen:
1.Solver settles partially fillable order for some sell amount.
2. Solver settles FOK order for some some sell amount.
3. User doesn't have enough balances for both fills and settlement fails.